File: minicraft-java/src/main/java/com/minicraft/world/ChunkMesher.java
package com.minicraft.world;

import com.minicraft.render.Tiles;

import static com.minicraft.world.Chunk.*;

/** Builds opaque + transparent meshes for a chunk (CPU-only; thread-safe). */
public final class ChunkMesher {

    public static final class Neighbors {
        public Chunk west, east, north, south;
    }

    // 6 faces: normal + base corner + in-plane du/dv unit steps.
    private static final int[][] FACE = {
        {  1, 0, 0, 1,0,0, 0,0,1, 0,1,0 },  // +X
        { -1, 0, 0, 0,0,1, 0,0,-1,0,1,0 },  // -X
        {  0, 1, 0, 0,1,1, 1,0,0, 0,0,-1},  // +Y
        {  0,-1, 0, 0,0,0, 1,0,0, 0,0,1 },  // -Y
        {  0, 0, 1, 1,0,1, -1,0,0,0,1,0 },  // +Z
        {  0, 0,-1, 0,0,0, 1,0,0, 0,1,0 },  // -Z
    };
    private static final float[] FACE_SHADE = {0.78f, 0.78f, 1.0f, 0.55f, 0.65f, 0.65f};
    private static final float[] AO_LUT = {0.45f, 0.62f, 0.80f, 1.0f};

    private final Chunk c;
    private final Neighbors nb;

    private ChunkMesher(Chunk c, Neighbors nb) { this.c = c; this.nb = nb; }

    public static void build(Chunk c, Neighbors nb) { new ChunkMesher(c, nb).run(); }

    private int blockAt(int x, int y, int z) {
        if (y < 0 || y >= SY) return Blocks.AIR;
        boolean xo = x < 0 || x >= SX, zo = z < 0 || z >= SZ;
        if (xo && zo) return Blocks.AIR;
        if (!xo && !zo) return c.get(x, y, z);
        if (xo) {
            Chunk n = x < 0 ? nb.west : nb.east;
            return n == null ? Blocks.AIR : n.get((x + SX) % SX, y, z);
        }
        Chunk n = z < 0 ? nb.north : nb.south;
        return n == null ? Blocks.AIR : n.get(x, y, (z + SZ) % SZ);
    }

    private boolean opaqueAt(int x, int y, int z) { return Blocks.isOpaque(blockAt(x, y, z)); }

    private int[] lightAt(int x, int y, int z) {
        if (y < 0) return new int[]{0, 0};
        if (y >= SY) return new int[]{15, 0};
        boolean xo = x < 0 || x >= SX, zo = z < 0 || z >= SZ;
        if (xo && zo) return new int[]{15, 0};
        Chunk n = c; int lx = x, lz = z;
        if (xo) { n = x < 0 ? nb.west : nb.east; lx = (x + SX) % SX; }
        else if (zo) { n = z < 0 ? nb.north : nb.south; lz = (z + SZ) % SZ; }
        if (n == null) return new int[]{15, 0};
        return new int[]{n.sky(lx, y, lz), n.block(lx, y, lz)};
    }

    private static boolean faceVisible(int self, int neighbour) {
        if (neighbour == Blocks.AIR) return true;
        if (Blocks.isOpaque(neighbour)) return false;
        return neighbour != self;
    }

    private static float aoLevel(boolean s1, boolean s2, boolean corner) {
        if (s1 && s2) return 0;
        int n = (s1 ? 1 : 0) + (s2 ? 1 : 0) + (corner ? 1 : 0);
        return 3 - n;
    }

    private static void atlasUV(int tile, float s, float t, float[] out) {
        int col = tile % Tiles.ATLAS_TILES, row = tile / Tiles.ATLAS_TILES;
        float inv = 1f / Tiles.ATLAS_PX;
        float inset = 0.5f * inv;
        float ts = Tiles.TILE_PX * inv;
        out[0] = col * ts + inset + s * (ts - 2 * inset);
        out[1] = row * ts + inset + t * (ts - 2 * inset);
    }

    private void run() {
        FloatBuf opaque = new FloatBuf(4096);
        FloatBuf transp = new FloatBuf(1024);
        float[] uv = new float[2], uv0 = new float[2], uv1 = new float[2];

        for (int y = 0; y < SY; y++)
            for (int z = 0; z < SZ; z++)
                for (int x = 0; x < SX; x++) {
                    int id = c.get(x, y, z);
                    if (id == Blocks.AIR) continue;
                    Blocks.Def def = Blocks.get(id);
                    boolean toTransp = def.render != Blocks.Render.SOLID;
                    FloatBuf out = toTransp ? transp : opaque;

                    if (def.render == Blocks.Render.CROSS) {
                        int[] l = lightAt(x, y, z);
                        float fsky = l[0] / 15f, fblk = l[1] / 15f;
                        atlasUV(def.texSide, 0, 1, uv0);
                        atlasUV(def.texSide, 1, 0, uv1);
                        float p = 0.146f;
                        emitCross(out, x + p, z + p, x + 1 - p, z + 1 - p, y, uv0, uv1, fsky, fblk);
                        emitCross(out, x + 1 - p, z + p, x + p, z + 1 - p, y, uv0, uv1, fsky, fblk);
                        continue;
                    }

                    for (int f = 0; f < 6; f++) {
                        int[] face = unpack(f);
                        int nbBlock = blockAt(x + face[0], y + face[1], z + face[2]);
                        if (!faceVisible(id, nbBlock)) continue;

                        int tile = f == 2 ? def.texTop : f == 3 ? def.texBottom : def.texSide;
                        float shade = FACE_SHADE[f];
                        boolean waterTop = def.render == Blocks.Render.LIQUID && f == 2;

                        float[] vx = new float[4 * 11];
                        float[] ao = new float[4];
                        for (int corner = 0; corner < 4; corner++) {
                            int cu = (corner == 1 || corner == 2) ? 1 : 0;
                            int cv = corner >= 2 ? 1 : 0;
                            float px = x + face[3] + cu * face[6] + cv * face[9];
                            float py = y + face[4] + cu * face[7] + cv * face[10];
                            float pz = z + face[5] + cu * face[8] + cv * face[11];
                            if (waterTop) py -= 0.12f;

                            int suX = cu != 0 ? face[6] : -face[6];
                            int suY = cu != 0 ? face[7] : -face[7];
                            int suZ = cu != 0 ? face[8] : -face[8];
                            int svX = cv != 0 ? face[9] : -face[9];
                            int svY = cv != 0 ? face[10] : -face[10];
                            int svZ = cv != 0 ? face[11] : -face[11];
                            int bx = x + face[0], by = y + face[1], bz = z + face[2];
                            boolean s1 = opaqueAt(bx + suX, by + suY, bz + suZ);
                            boolean s2 = opaqueAt(bx + svX, by + svY, bz + svZ);
                            boolean cc = opaqueAt(bx + suX + svX, by + suY + svY, bz + suZ + svZ);
                            ao[corner] = aoLevel(s1, s2, cc);

                            int accSky = 0, accBlk = 0, cnt = 0;
                            int[][] samples = {{0,0,0},{suX,suY,suZ},{svX,svY,svZ},{suX+svX,suY+svY,suZ+svZ}};
                            for (int[] s : samples) {
                                if (opaqueAt(bx + s[0], by + s[1], bz + s[2])) continue;
                                int[] ll = lightAt(bx + s[0], by + s[1], bz + s[2]);
                                accSky += ll[0]; accBlk += ll[1]; cnt++;
                            }
                            if (cnt == 0) { int[] ll = lightAt(bx, by, bz); accSky = ll[0]; accBlk = ll[1]; cnt = 1; }
                            float fsky = (accSky / (float) cnt) / 15f;
                            float fblk = (accBlk / (float) cnt) / 15f;

                            float sCoord = cu, tCoord = cv != 0 ? 0f : 1f;
                            atlasUV(tile, sCoord, tCoord, uv);
                            int o = corner * 11;
                            vx[o] = px; vx[o+1] = py; vx[o+2] = pz; vx[o+3] = uv[0]; vx[o+4] = uv[1];
                            vx[o+5] = face[0]; vx[o+6] = face[1]; vx[o+7] = face[2];
                            vx[o+8] = AO_LUT[(int) ao[corner]] * shade; vx[o+9] = fsky; vx[o+10] = fblk;
                        }

                        boolean flip = (ao[0] + ao[2]) < (ao[1] + ao[3]);
                        if (flip) emitTri(out, vx, 1, 2, 3, 1, 3, 0);
                        else emitTri(out, vx, 0, 1, 2, 0, 2, 3);
                    }
                }

        c.meshOpaque = opaque.toArray();
        c.meshTransparent = transp.toArray();
    }

    private int[] unpack(int f) { return FACE[f]; }

    private static void emitTri(FloatBuf out, float[] vx, int... order) {
        for (int idx : order) {
            int o = idx * 11;
            out.vertex(vx[o], vx[o+1], vx[o+2], vx[o+3], vx[o+4], vx[o+5], vx[o+6], vx[o+7], vx[o+8], vx[o+9], vx[o+10]);
        }
    }

    private static void emitCross(FloatBuf out, float ax, float az, float bx, float bz, int y,
                                  float[] uv0, float[] uv1, float fsky, float fblk) {
        float top = y + 1f;
        // a..b at bottom, up to top.
        out.vertex(ax, y, az, uv0[0], uv1[1], 0,1,0, 1f, fsky, fblk);
        out.vertex(bx, y, bz, uv1[0], uv1[1], 0,1,0, 1f, fsky, fblk);
        out.vertex(bx, top, bz, uv1[0], uv0[1], 0,1,0, 1f, fsky, fblk);
        out.vertex(ax, y, az, uv0[0], uv1[1], 0,1,0, 1f, fsky, fblk);
        out.vertex(bx, top, bz, uv1[0], uv0[1], 0,1,0, 1f, fsky, fblk);
        out.vertex(ax, top, az, uv0[0], uv0[1], 0,1,0, 1f, fsky, fblk);
    }
}

File: src/World/ChunkMesher.cpp
#include "World/ChunkMesher.h"
#include "World/Chunk.h"
#include "World/Block.h"
#include "Render/TileIds.h"
#include <array>
#include <cmath>

namespace mc {

namespace {

// 6 cube faces. base = corner where (u,v)=(0,0); du,dv = in-plane unit steps.
struct Face {
    int nx, ny, nz;
    int base[3];
    int du[3];
    int dv[3];
};

const Face FACES[6] = {
    {  1, 0, 0, {1,0,0}, {0,0,1}, {0,1,0} }, // +X
    { -1, 0, 0, {0,0,1}, {0,0,-1},{0,1,0} }, // -X
    {  0, 1, 0, {0,1,1}, {1,0,0}, {0,0,-1}}, // +Y (top)
    {  0,-1, 0, {0,0,0}, {1,0,0}, {0,0,1} }, // -Y (bottom)
    {  0, 0, 1, {1,0,1}, {-1,0,0},{0,1,0} }, // +Z
    {  0, 0,-1, {0,0,0}, {1,0,0}, {0,1,0} }, // -Z
};

// Per-face brightness so flat-lit faces still read as 3D (Minecraft-style).
const float FACE_SHADE[6] = { 0.78f, 0.78f, 1.0f, 0.55f, 0.65f, 0.65f };

float aoLevel(bool s1, bool s2, bool corner) {
    if (s1 && s2) return 0.0f;
    int n = (s1 ? 1 : 0) + (s2 ? 1 : 0) + (corner ? 1 : 0);
    return float(3 - n);
}
float aoToBrightness(float ao) {
    static const float lut[4] = {0.45f, 0.62f, 0.80f, 1.0f};
    int i = (int)ao;
    if (i < 0) i = 0; if (i > 3) i = 3;
    return lut[i];
}

// Resolves a local coordinate possibly outside the centre chunk to the right
// neighbour. Returns air when a needed (esp. diagonal) neighbour is missing.
struct Access {
    Chunk* c;
    const ChunkNeighbors& nb;

    BlockId block(int x, int y, int z) const {
        if (y < 0 || y >= CHUNK_SY) return BLOCK_AIR;
        bool xo = (x < 0 || x >= CHUNK_SX);
        bool zo = (z < 0 || z >= CHUNK_SZ);
        if (xo && zo) return BLOCK_AIR;
        if (!xo && !zo) return c->get(x, y, z);
        if (xo) {
            Chunk* n = (x < 0) ? nb.west : nb.east;
            if (!n) return BLOCK_AIR;
            return n->get((x + CHUNK_SX) % CHUNK_SX, y, z);
        }
        Chunk* n = (z < 0) ? nb.north : nb.south;
        if (!n) return BLOCK_AIR;
        return n->get(x, y, (z + CHUNK_SZ) % CHUNK_SZ);
    }
    void light(int x, int y, int z, int& sky, int& blk) const {
        if (y < 0) { sky = 0; blk = 0; return; }
        if (y >= CHUNK_SY) { sky = 15; blk = 0; return; }
        bool xo = (x < 0 || x >= CHUNK_SX);
        bool zo = (z < 0 || z >= CHUNK_SZ);
        if (xo && zo) { sky = 15; blk = 0; return; }
        Chunk* n = c;
        int lx = x, lz = z;
        if (xo) { n = (x < 0) ? nb.west : nb.east; lx = (x + CHUNK_SX) % CHUNK_SX; }
        else if (zo) { n = (z < 0) ? nb.north : nb.south; lz = (z + CHUNK_SZ) % CHUNK_SZ; }
        if (!n) { sky = 15; blk = 0; return; }
        sky = n->skyLight(lx, y, lz);
        blk = n->blockLight(lx, y, lz);
    }
    bool opaque(int x, int y, int z) const { return Blocks::isOpaque(block(x, y, z)); }
};

// Whether a face between `self` and `neighbour` should be emitted.
bool faceVisible(BlockId self, BlockId nb) {
    if (nb == BLOCK_AIR) return true;
    if (Blocks::isOpaque(nb)) return false;
    if (nb == self) return false;       // hide internal faces of same translucent type
    return true;
}

void atlasUV(uint16_t tile, float s, float t, float& u, float& v) {
    int col = tile % ATLAS_TILES;
    int row = tile / ATLAS_TILES;
    const float inv = 1.0f / ATLAS_PX;
    const float inset = 0.5f * inv; // avoid mipmap bleed
    float tileSize = float(TILE_PX) * inv;
    u = col * tileSize + inset + s * (tileSize - 2 * inset);
    v = row * tileSize + inset + t * (tileSize - 2 * inset);
}

uint16_t faceTile(const BlockDef& def, int faceIdx) {
    if (faceIdx == 2) return def.texTop;     // +Y
    if (faceIdx == 3) return def.texBottom;  // -Y
    return def.texSide;
}

} // namespace

void buildChunkMesh(Chunk& chunk, const ChunkNeighbors& nb) {
    chunk.mesh.opaque.clear();
    chunk.mesh.transparent.clear();
    Access acc{&chunk, nb};

    for (int y = 0; y < CHUNK_SY; ++y)
        for (int z = 0; z < CHUNK_SZ; ++z)
            for (int x = 0; x < CHUNK_SX; ++x) {
                BlockId id = chunk.get(x, y, z);
                if (id == BLOCK_AIR) continue;
                const BlockDef& def = Blocks::get(id);
                RenderKind kind = def.render;

                bool transparentMesh = (kind != RenderKind::Solid);
                auto& out = transparentMesh ? chunk.mesh.transparent : chunk.mesh.opaque;

                if (kind == RenderKind::Cross) {
                    // Two crossed quads. Light from this cell.
                    int sky, blk; acc.light(x, y, z, sky, blk);
                    float fsky = sky / 15.0f, fblk = blk / 15.0f;
                    float u0, v0, u1, v1;
                    atlasUV(def.texSide, 0, 1, u0, v1);
                    atlasUV(def.texSide, 1, 0, u1, v0);
                    const float p = 0.146f; // inset so quads fit the cell diagonal
                    struct QV { float x, z; float u; };
                    // Quad A: (p,p)->(1-p,1-p)
                    glm::vec3 A0(x + p, y, z + p), A1(x + 1 - p, y, z + 1 - p);
                    glm::vec3 B0(x + 1 - p, y, z + p), B1(x + p, y, z + 1 - p);
                    auto emitCross = [&](glm::vec3 a, glm::vec3 b) {
                        float top = y + 1.0f;
                        Vertex va{a.x, a.y, a.z, u0, v1, 0,1,0, 1.0f, fsky, fblk};
                        Vertex vb{b.x, b.y, b.z, u1, v1, 0,1,0, 1.0f, fsky, fblk};
                        Vertex vc{b.x, top, b.z, u1, v0, 0,1,0, 1.0f, fsky, fblk};
                        Vertex vd{a.x, top, a.z, u0, v0, 0,1,0, 1.0f, fsky, fblk};
                        out.push_back(va); out.push_back(vb); out.push_back(vc);
                        out.push_back(va); out.push_back(vc); out.push_back(vd);
                    };
                    emitCross(A0, A1);
                    emitCross(B0, B1);
                    continue;
                }

                for (int f = 0; f < 6; ++f) {
                    const Face& face = FACES[f];
                    int nxp = x + face.nx, nyp = y + face.ny, nzp = z + face.nz;
                    BlockId neighbour = acc.block(nxp, nyp, nzp);
                    if (!faceVisible(id, neighbour)) continue;

                    uint16_t tile = faceTile(def, f);
                    float shade = FACE_SHADE[f];

                    // Liquid surface: lower the top face when open above.
                    float topDrop = 0.0f;
                    bool isWaterTop = (kind == RenderKind::Liquid && f == 2);
                    if (isWaterTop) topDrop = 0.12f;

                    Vertex quad[4];
                    float aoVals[4];
                    for (int corner = 0; corner < 4; ++corner) {
                        int cu = (corner == 1 || corner == 2) ? 1 : 0;
                        int cv = (corner >= 2) ? 1 : 0;
                        float px = x + face.base[0] + cu * face.du[0] + cv * face.dv[0];
                        float py = y + face.base[1] + cu * face.du[1] + cv * face.dv[1];
                        float pz = z + face.base[2] + cu * face.du[2] + cv * face.dv[2];
                        if (isWaterTop) py -= topDrop;

                        // AO sampling around this corner in the neighbour plane.
                        int suX = (cu ? face.du[0] : -face.du[0]);
                        int suY = (cu ? face.du[1] : -face.du[1]);
                        int suZ = (cu ? face.du[2] : -face.du[2]);
                        int svX = (cv ? face.dv[0] : -face.dv[0]);
                        int svY = (cv ? face.dv[1] : -face.dv[1]);
                        int svZ = (cv ? face.dv[2] : -face.dv[2]);
                        int bx = x + face.nx, by = y + face.ny, bz = z + face.nz;
                        bool s1 = acc.opaque(bx + suX, by + suY, bz + suZ);
                        bool s2 = acc.opaque(bx + svX, by + svY, bz + svZ);
                        bool cc = acc.opaque(bx + suX + svX, by + suY + svY, bz + suZ + svZ);
                        float ao = aoLevel(s1, s2, cc);
                        aoVals[corner] = ao;

                        // Smooth light: average non-opaque cells around the corner.
                        int accSky = 0, accBlk = 0, cnt = 0;
                        auto sample = [&](int ox, int oy, int oz) {
                            if (acc.opaque(bx + ox, by + oy, bz + oz)) return;
                            int s, b; acc.light(bx + ox, by + oy, bz + oz, s, b);
                            accSky += s; accBlk += b; ++cnt;
                        };
                        sample(0, 0, 0);
                        sample(suX, suY, suZ);
                        sample(svX, svY, svZ);
                        sample(suX + svX, suY + svY, suZ + svZ);
                        if (cnt == 0) { acc.light(bx, by, bz, accSky, accBlk); cnt = 1; }
                        float fsky = (accSky / float(cnt)) / 15.0f;
                        float fblk = (accBlk / float(cnt)) / 15.0f;

                        float s = float(cu), t = (cv ? 0.0f : 1.0f);
                        float uu, vv; atlasUV(tile, s, t, uu, vv);

                        quad[corner] = Vertex{
                            px, py, pz, uu, vv,
                            float(face.nx), float(face.ny), float(face.nz),
                            aoToBrightness(ao) * shade, fsky, fblk};
                    }

                    // Flip the quad's diagonal to avoid AO interpolation seams.
                    bool flip = (aoVals[0] + aoVals[2]) < (aoVals[1] + aoVals[3]);
                    if (flip) {
                        out.push_back(quad[1]); out.push_back(quad[2]); out.push_back(quad[3]);
                        out.push_back(quad[1]); out.push_back(quad[3]); out.push_back(quad[0]);
                    } else {
                        out.push_back(quad[0]); out.push_back(quad[1]); out.push_back(quad[2]);
                        out.push_back(quad[0]); out.push_back(quad[2]); out.push_back(quad[3]);
                    }
                }
            }
}

} // namespace mc
